homeassistant/components/vicare/number.py

```diff
@@ -0,0 +1,194 @@
+        value_setter=lambda api, slope: api.setHeatingCurve(
+            api.getHeatingCurveShift(), slope
+        ),
+        native_min_value=0.2,
+        native_max_value=3.5,
+        native_step=0.1,
+    ),
+)
+
+
+def _build_entity(
+    name: str,
+    vicare_api: PyViCareHeatingDeviceWithComponent,
+    device_config: PyViCareDeviceConfig,
+    entity_description: ViCareNumberEntityDescription,
+    hass: HomeAssistant,
+) -> ViCareNumber | None:
+    """Create a ViCare number entity."""
+    _LOGGER.debug("Found device %s", name)
+    if is_supported(name, entity_description, vicare_api):
+        return ViCareNumber(
+            name,
+            vicare_api,
+            device_config,
+            entity_description,
+            hass,
+        )
+    return None
+
+
+async def _entities_from_descriptions(
+    hass: HomeAssistant,
+    entities: list[ViCareNumber],
+    sensor_descriptions: tuple[ViCareNumberEntityDescription, ...],
+    iterables: list[PyViCareHeatingDeviceWithComponent],
+    config_entry: ConfigEntry,
+) -> None:
+    """Create entities from descriptions and list of burners/circuits."""
+    for description in sensor_descriptions:
+        for current in iterables:
+            suffix = ""
+            if len(iterables) > 1:
+                suffix = f" {current.id}"
+            entity = await hass.async_add_executor_job(
+                _build_entity,
+                f"{description.name}{suffix}",
+                current,
+                hass.data[DOMAIN][config_entry.entry_id][VICARE_DEVICE_CONFIG],
+                description,
+                hass,
+            )
+            if entity is not None:
+                entities.append(entity)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Create the ViCare sensor devices."""
+    api = hass.data[DOMAIN][config_entry.entry_id][VICARE_API]
+
+    entities: list[ViCareNumber] = []
+    try:
+        await _entities_from_descriptions(
+            hass, entities, CIRCUIT_SENSORS, api.circuits, config_entry
+        )
+    except PyViCareNotSupportedFeatureError:
+        _LOGGER.debug("No circuits found")
+
+    async_add_entities(entities)
+
+
+class ViCareNumber(ViCareEntity, NumberEntity):
+    """Representation of a ViCare sensor."""
+
+    entity_description: ViCareNumberEntityDescription
+
+    def __init__(
+        self,
+        name: str,
+        api: PyViCareHeatingDeviceWithComponent,
+        device_config: PyViCareDeviceConfig,
+        description: ViCareNumberEntityDescription,
+        hass: HomeAssistant,
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__(device_config)
+        self.entity_description = description
+        self._attr_name = name
+        self._api = api
+        self._device_config = device_config
+        self._hass = hass
```

Comment:
Is there a better way to get `hass` than handing it in the entity here?

Reply:
self.hass will be set on the entity once its been async_added_to_hass